bundles/org.openhab.core/src/test/java/org/openhab/core/cache/ByteArrayFileCacheTest.java

```diff
@@ -0,0 +1,203 @@
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.openhab.core.OpenHAB;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test class for the {@link ByteArrayFileCache} class.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class ByteArrayFileCacheTest {
+
+    private static final String SERVICE_PID = "org.openhab.core.test";
+
+    private static final File USERDATA_FOLDER = new File(OpenHAB.getUserDataFolder());
+    private static final File CACHE_FOLDER = new File(USERDATA_FOLDER, ByteArrayFileCache.CACHE_FOLDER_NAME);
+    private static final File SERVICE_CACHE_FOLDER = new File(CACHE_FOLDER, SERVICE_PID);
+
+    private static final String MP3_FILE_NAME = SERVICE_CACHE_FOLDER.getAbsolutePath() + "doorbell.mp3";
+    private static final String TXT_FILE_NAME = SERVICE_CACHE_FOLDER.getAbsolutePath() + "doorbell.txt";
+
+    private static @Nullable File txt_file;
+
+    private final Logger logger = LoggerFactory.getLogger(ByteArrayFileCacheTest.class);
+
+    private @NonNullByDefault({}) ByteArrayFileCache subject;
+
+    @BeforeAll
+    public static void init() throws IOException {
+        // create temporary file
+        txt_file = createTempFile();
+    }
+
+    @BeforeEach
+    public void setUp() {
+        subject = new ByteArrayFileCache(SERVICE_PID);
+    }
+
+    @AfterEach
+    public void tearDown() {
+        // delete all files
+        subject.clear();
+    }
+
+    @AfterAll
+    public static void cleanUp() {
+        // delete all folders
+        SERVICE_CACHE_FOLDER.delete();
+        CACHE_FOLDER.delete();
+        USERDATA_FOLDER.delete();
+    }
+
+    @Test
+    public void testGetFileExtension() {
+        assertThat(subject.getFileExtension("/var/log/openhab2/"), is(nullValue()));
+        assertThat(subject.getFileExtension("/var/log/foo.bar/"), is(nullValue()));
+        assertThat(subject.getFileExtension("doorbell.mp3"), is(equalTo("mp3")));
+        assertThat(subject.getFileExtension("/tmp/doorbell.mp3"), is(equalTo("mp3")));
+        assertThat(subject.getFileExtension(MP3_FILE_NAME), is(equalTo("mp3")));
+        assertThat(subject.getFileExtension(TXT_FILE_NAME), is(equalTo("txt")));
+        assertThat(subject.getFileExtension("/var/log/openhab2/.."), is(""));
+        assertThat(subject.getFileExtension(".hidden"), is(equalTo("hidden")));
+        assertThat(subject.getFileExtension("C:\\Program Files (x86)\\java\\bin\\javaw.exe"), is(equalTo("exe")));
+        assertThat(subject.getFileExtension("https://www.youtube.com/watch?v=qYrpPrLY868"), is(nullValue()));
```

Comment:
Is this testing, what is implemented? None of the tests go into the branch to remove the `?` at the end of the extension.
What about:

- https://www.youtube.com/watch?v=Test.With.Dots
- https://host/test.xlsx?cache=false

Reply:
:+1: Good point.

First one fails:
```
assertThat(subject.getFileExtension("https://www.youtube.com/watch?v=Test.With.Dots"), is(nullValue()));
```
To got it sorted I think I have to strip the `?` and the content after it from the file name first.

Second one is working:
```
assertThat(subject.getFileExtension("https://host/test.xlsx?cache=false"), is("xlsx"));
```

I can remember a situation where someone who uses Kodi (player) combined with Plex (media server) in a local network sent me an error in which an url like the following got interpreted wrong because of the local ip address:

- http://127.0.0.1:8080/image/image%3A%2F%2Fhttp%253a%252f%252f127.0.0.1%253a32400%252fphoto%252f%253a%252ftranscode%253fwidth%253d1920%2526height%253d1920%2526minSize%253d1%2526upscale%253d0%2526url%253d%252flibrary%252fmetadata%252f1896%252fthumb%252f1569782004%2526X-Plex-Token%253dXScJLJbUdcybNXFyHLuv

I never found a solution for it. The extension will result in `1%253a32400%252fphoto%252f%253a%252ftranscode%253fwidth%253d1920%2526height%253d1920%2526minSize%253d1%2526upscale%253d0%2526url%253d%252flibrary%252fmetadata%252f1896%252fthumb%252f1569782004%2526X-Plex-Token%253dXScJLJbUdcybNXFyHLuv`. But I think it is a very rare situation.

Reply:
I saw a problem with the first one.
But we should add the tests to be sure, that it keeps working even after firther development :-)

Reply:
Already done in https://github.com/openhab/openhab-core/pull/1723/commits/9769174086818da34834ad4d8a575d2a0622f689.